File: 2ndSemester/AlgorithimsAndDataStructuresII/Important_Data_Structures/WithArrays/Java_Implementation/Queue.java
class Queue{
	int[] queue;
	int first, last;
	Queue(){
		this(7);
	}
	Queue(int size){
		queue = new int[size + 1];
		first = last = 0;
	}
	void insert(int x) throws Exception{
		if(last + 1 % queue.length == 0){
			throw new Exception("Full queue");
		}
		queue[last] = x;
		last = (last + 1) % queue.length;
	}
	
	int remove()  throws Exception{
		int removed;
		if(first == last){
			throw new Exception("Empty list");
		}
		removed = queue[first];
		first = (first + 1) % queue.length;
		return removed;
	}
	void show(){
		int i = first;
		System.out.print("[");
		while(i != last){
			System.out.print(queue[i] + ",");
			i = (i  + 1) % queue.length;
		}
		System.out.println("]");
	}
}

File: 2ndSemester/AlgorithimsAndDataStructuresII/U05-AbstractDataTypes_linear/c/queue.c
#include <stdio.h>
#include <stdlib.h>
#define MAX_LENGTH 6 //uma posicao do array nao sera usada

int first, last;

void start(){
  int first = last = 0;
}

void enqueue(int x, int *queue){
  if(((last + 1) % MAX_LENGTH) == first){
    printf("A fila esta cheia");
  }
  else{
    queue[last] = x;
    last = (last + 1) % MAX_LENGTH;
  }
}

int dequeue(int *queue){
  int removed = 0;
  if(first != last){
    removed = queue[first];
    first = (first + 1) % MAX_LENGTH;
  }
  return removed; 
}

void display(int *queue){
  int i = first;
  printf("Sua fila:\n");
  while(i != last){
    printf("%d - ", queue[i]);
    i = (i + 1) % MAX_LENGTH;
  }
  printf("\n");
}

void enqueueTest(int *queue){
  for(int i = 0; i < MAX_LENGTH - 1; i ++)
    enqueue(i, queue);
  display(queue);
}

void dequeueTest(int *queue){
  for(int i = 0; i < 3; i++)
    printf("\t- %d foi removido da fila\n", dequeue(queue));
  display(queue);
}

int main(){
  start();
  int queue[MAX_LENGTH];
  enqueueTest(queue);
  dequeueTest(queue);
  return 0;
}
